python/src/robyn/modeling/feature_engineering.py

```diff
@@ -0,0 +1,207 @@
+from typing import List, Optional, Dict, Any, Tuple
+import pandas as pd
+import numpy as np
+from dataclasses import dataclass
+
+from robyn.data.entities.enums import (
+    DependentVarType,
+    AdstockType,
+    SaturationType,
+    ProphetVariableType,
+    PaidMediaSigns,
+    OrganicSigns,
+    ContextSigns,
+    ProphetSigns,
+    CalibrationScope,
+)
+
+from robyn.data.entities.calibration_input import CalibrationInput, ChannelCalibrationData
+from robyn.data.entities.hyperparameters import Hyperparameters, ChannelHyperparameters
+from robyn.data.entities.mmmdata import MMMData
+
+
+@dataclass
+class FeaturizedMMMData:
+    """
+    A dataclass to hold the output of the feature engineering process.
+
+    Attributes:
+        dt_mod (pd.DataFrame): The transformed data after feature engineering.
+        dt_modRollWind (pd.DataFrame): The rolling window data used in the modeling process.
+        modNLS (Dict[str, Any]): Results of the non-linear models, including fitted models, plots, and predictions.
+    """
+
+    dt_mod: pd.DataFrame
+    dt_modRollWind: pd.DataFrame
+    modNLS: Dict[str, Any]
+
+
+class FeatureEngineering:
+    """
+    A class to perform feature engineering for Marketing Mix Modeling (MMM).
+
+    This class handles various aspects of feature engineering, including data preparation,
+    rolling window creation, media cost factor calculation, non-linear modeling,
+    and prophet decomposition.
+
+    Attributes:
+        mmm_data (MMMData): The input data and specifications for the MMM.
+        hyperparameters (Hyperparameters): The hyperparameters for the model.
+    """
+
+    def __init__(self, mmm_data: MMMData, hyperparameters: Hyperparameters):
+        """
+        Initialize the FeatureEngineering class.
+
+        Args:
+            mmm_data (MMMData): The input data and specifications for the MMM.
+            hyperparameters (Hyperparameters): The hyperparameters for the model.
+        """
+        self.mmm_data = mmm_data
+        self.hyperparameters = hyperparameters
+
+    def perform_feature_engineering(self, quiet: bool = False) -> FeaturizedMMMData:
+        """
+        Perform the feature engineering process.
+
+        This method orchestrates the entire feature engineering pipeline, including
+        data preparation, rolling window creation, media cost factor calculation,
+        model running, and prophet decomposition (if applicable).
+
+        Args:
+            quiet (bool, optional): If True, suppresses print statements. Defaults to False.
+
+        Returns:
+            FeaturizedMMMData: The output of the feature engineering process.
+        """
+        if not quiet:
```

Comment:
No quiet flags or no print statements anywhere. We will be using logging framework that is already added with file handler (prints into log file ) and console handler (prints onto console). Console handler can be updated to remove datetime prefix if we want it cleaner.

Reply:
removed - will move integration with logging framework as part of implementation PR.